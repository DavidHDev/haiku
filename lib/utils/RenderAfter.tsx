import { type ReactNode, useEffect, useState } from 'react';

type Props = {
  delay?: number;
  children: ReactNode;
};

export const RenderAfter = ({ delay = 1000, children }: Props) => {
  const [ready, setReady] = useState(false);

  useEffect(() => {
    const timer = setTimeout(() => setReady(true), delay);
    return () => clearTimeout(timer);
  }, [ready]);

  return ready ? children : null;
};
