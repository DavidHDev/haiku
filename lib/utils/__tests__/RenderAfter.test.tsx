import { afterEach, beforeEach, describe, expect, it, vi } from 'vitest';
import { act, render, screen } from '@testing-library/react';
import { RenderAfter } from '../RenderAfter';

describe('RederAfter', () => {
  const text = 'Delayed Content';

  beforeEach(() => {
    vi.useFakeTimers({ shouldAdvanceTime: true });
  });

  afterEach(() => {
    vi.runOnlyPendingTimers();
    vi.useRealTimers();
  });

  it('renders the RenderAfter component correctly after delay', async () => {
    const { asFragment } = render(<RenderAfter>{text}</RenderAfter>);
    act(() => vi.advanceTimersByTime(1000));
    expect(asFragment()).toMatchSnapshot();
    expect(screen.getByText(text)).toBeInTheDocument();
  });

  it('does not render children initially', () => {
    render(<RenderAfter>{text}</RenderAfter>);
    expect(screen.queryByText(text)).not.toBeInTheDocument();
  });

  it('renders children with different delay', async () => {
    const delay = 2000;
    render(<RenderAfter delay={delay}>{text}</RenderAfter>);
    expect(screen.queryByText(text)).not.toBeInTheDocument();
    act(() => vi.advanceTimersByTime(delay));
    expect(screen.getByText(text)).toBeInTheDocument();
  });
});
