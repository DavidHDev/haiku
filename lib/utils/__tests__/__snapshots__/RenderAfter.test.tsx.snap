// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`RederAfter > renders the RenderAfter component correctly after delay 1`] = `
<DocumentFragment>
  Delayed Content
</DocumentFragment>
`;
